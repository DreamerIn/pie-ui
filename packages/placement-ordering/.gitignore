.idea
node_modules
docs/demo/.pie
lib