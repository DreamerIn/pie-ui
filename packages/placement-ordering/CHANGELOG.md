# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.1"></a>
## 2.0.1 (2018-03-06)


### Bug Fixes

* **package:** fix repo name ([c96be96](https://github.com/pie-framework/pie-ui/commit/c96be96))
* **repository:** fix repository ([ed63b70](https://github.com/pie-framework/pie-ui/commit/ed63b70))
