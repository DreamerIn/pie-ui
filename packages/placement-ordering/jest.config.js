module.exports = {
  verbose: true,
  testEnvironment: 'node',
  setupFiles: ['./jest.setup.js'],
  testRegex: '/__tests__/.*test.jsx?$'
};
