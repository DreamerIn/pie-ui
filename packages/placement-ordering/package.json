{
  "name": "@pie-ui/placement-ordering",
  "version": "4.4.3",
  "main": "lib/index.js",
  "description": "A pie placement component",
  "keywords": [
    "pie"
  ],
  "repository": "pie-framework/pie-ui",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prepack": "../../scripts/build changelog --scope placement-ordering"
  },
  "dependencies": {
    "@material-ui/core": "^3.9.2",
    "@pie-lib/correct-answer-toggle": "^2.1.2",
    "@pie-lib/drag": "^1.1.4",
    "@pie-lib/math-rendering": "^1.2.1",
    "@pie-lib/render-ui": "^4.3.1",
    "debug": "^3.1.0",
    "decimal.js": "^10.0.0",
    "lodash": "^4.17.10",
    "prop-types": "^15.6.1",
    "react": "^16.8.1",
    "react-dnd": "^3.0.0",
    "react-dnd-html5-backend": "^2.6.0",
    "react-dom": "^16.8.1",
    "react-jss": "^8.4.0",
    "react-transition-group": "^2.3.1"
  },
  "devDependencies": {
    "react-dnd-test-backend": "^2.6.0"
  },
  "gitHead": "25b2169d13da6abaaaa962e42f3f6d6ab400ec8b"
}
