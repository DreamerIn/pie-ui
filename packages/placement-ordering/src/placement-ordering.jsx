import { HorizontalTiler, VerticalTiler } from './tiler';
import { buildState, reducer } from './ordering';
import { Feedback } from '@pie-lib/render-ui';
import CorrectAnswerToggle from '@pie-lib/correct-answer-toggle';
import PropTypes from 'prop-types';
import React from 'react';
import cloneDeep from 'lodash/cloneDeep';
import compact from 'lodash/compact';
import debug from 'debug';
import isEmpty from 'lodash/isEmpty';
import uniqueId from 'lodash/uniqueId';
import { withStyles } from '@material-ui/core/styles';

const log = debug('pie-elements:placement-ordering');

export class PlacementOrdering extends React.Component {
  static propTypes = {
    onSessionChange: PropTypes.func.isRequired,
    model: PropTypes.object.isRequired,
    session: PropTypes.oneOfType([
      PropTypes.array.isRequired,
      PropTypes.object.isRequired,
    ]),
    classes: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      showingCorrect: false
    };

    this.instanceId = uniqueId();

    this.toggleCorrect = showingCorrect => {
      this.setState({ showingCorrect });
    };
  }

  componentDidMount() {
    this.initSessionIfNeeded(this.props);
  }

  componentWillReceiveProps(nextProps) {
    const newState = {};

    if (!nextProps.model.correctResponse) {
      newState.showingCorrect = false;
    }

    const includeTargetsChanged = nextProps.model.config.includeTargets !== this.props.model.config.includeTargets;
    const choicesNumberChanged = nextProps.model.choices.length !== this.props.model.choices.length;

    if (includeTargetsChanged || choicesNumberChanged) {
      this.initSessionIfNeeded(nextProps);
    }

    this.setState(newState);
  }

  initSessionIfNeeded(props) {
    const { model, session, onSessionChange } = props;
    const { config: newConfig } = model;

    const compactSessionValues = (session && compact(session.value)) || [];

    if (!newConfig.includeTargets && compactSessionValues.length !== model.choices.length) {
      log('[initSessionIfNeeded] initing session...', newConfig.includeTargets);
      const update = cloneDeep(session);
      update.value = model.choices.map(m => m.id);
      onSessionChange(update);
    } else if (newConfig.includeTargets) {
      const update = cloneDeep(session);

      delete update.value;

      onSessionChange(update);
    }
  }

  onDropChoice(target, source) {
    const { onSessionChange, session } = this.props;
    const from = this.ordering.tiles.find(
      t => t.id === source.id && t.type === source.type
    );
    const to = target;
    log('[onDropChoice] ', from, to);
    const update = reducer({ type: 'move', from, to }, this.ordering);
    const sessionUpdate = Object.assign({}, session, {
      value: update.response
    });
    onSessionChange(sessionUpdate);
  }

  onRemoveChoice(target) {
    const { onSessionChange, session } = this.props;
    log('[onRemoveChoice]', target);
    const update = reducer({ type: 'remove', target }, this.ordering);
    const sessionUpdate = Object.assign({}, session, {
      value: update.response
    });

    onSessionChange(sessionUpdate);
  }

  createOrdering = () => {
    const { model, session } = this.props;
    const { showingCorrect } = this.state;
    const config = model.config || {
        orientation: 'vertical',
        includeTargets: true
      };
    const { includeTargets } = config;

    return showingCorrect
      ? buildState(
        model.choices,
        model.correctResponse,
        model.correctResponse.map(id => ({ id, outcome: 'correct' })),
        { includeTargets },
        model.config.removeTile
      )
      : buildState(model.choices, session.value, model.outcomes, {
          includeTargets
        },
        model.config.removeTile
      );
  };

  render() {
    const { classes, model } = this.props;
    const showToggle =
      model.correctResponse && model.correctResponse.length > 0;
    const { showingCorrect } = this.state;
    const config = model.config || {
      orientation: 'vertical',
      includeTargets: true
    };
    const { orientation, includeTargets } = config;
    const vertical = orientation === 'vertical';

    this.ordering = this.createOrdering();

    const Tiler = vertical ? VerticalTiler : HorizontalTiler;

    return (
      <div className={classes.placementOrdering}>
        <CorrectAnswerToggle
          show={showToggle}
          toggled={this.state.showingCorrect}
          onToggle={this.toggleCorrect}
        />

        <div
          className={classes.prompt}
          dangerouslySetInnerHTML={{ __html: model.prompt }}
        />

        <Tiler
          instanceId={this.instanceId}
          choiceLabel={config.choiceLabel}
          targetLabel={config.targetLabel}
          tiles={this.ordering.tiles}
          disabled={model.disabled}
          addGuide={model.config.showOrdering}
          tileSize={model.config && model.config.tileSize}
          includeTargets={includeTargets}
          onDropChoice={this.onDropChoice.bind(this)}
          onRemoveChoice={this.onRemoveChoice.bind(this)}
        />

        <br />
        {!showingCorrect && (
          <Feedback correctness={model.correctness} feedback={model.feedback} />
        )}
      </div>
    );
  }
}

const styles = {
  placementOrdering: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    boxSizing: 'border-box'
  },
  prompt: {
    padding: '5px',
    paddingBottom: '15px'
  }
};

export default withStyles(styles)(PlacementOrdering);
