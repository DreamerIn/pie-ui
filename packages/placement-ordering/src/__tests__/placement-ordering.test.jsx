import { shallow } from 'enzyme';

import React from 'react';
import compact from 'lodash/compact';
import { PlacementOrdering, Choice } from '../placement-ordering';
import toJson from 'enzyme-to-json';
import CorrectAnswerToggle from '@pie-lib/correct-answer-toggle';

describe('PlacementOrdering', () => {
  let wrapper, model, session;

  const mkWrapper = (model, session) => {
    session = { value: [], ...session };
    model = { config: {}, ...model };

    return shallow(
      <PlacementOrdering
        model={model}
        session={session}
        classes={{
          placementOrdering: 'placementOrdering'
        }}
        onSessionChange={jest.fn()}
      />,
      {}
    );
  };

  beforeEach(() => {
    model = {
      choices: [
        {
          id: 'c1',
          label: 'C1'
        },
        {
          id: 'c2',
          label: 'C2'
        },
        {
          id: 'c3',
          label: 'C3'
        },
        {
          id: 'c4',
          label: 'C4'
        }
      ],
      config: {}
    };

    wrapper = mkWrapper(model);
    wrapper.setProps({
      onSessionChange: jest.fn(newSession => {
        wrapper.setState({ session: newSession })
      })
    });
  });

  describe('render', () => {
    it('snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('interaction', () => {
    it('dropping choices updates the order', () => {
      wrapper.instance().onDropChoice({ id: 'c4', type: 'choice'}, { id: 'c1', type: 'choice' });

      expect(wrapper.state('session').value).toEqual(['c4', 'c2', 'c3', 'c1']);
    });

    it('removing choices updates state', () => {
      session = { value: ['c1', 'c2', 'c3', 'c4'] };
      const valuesLength = session.value.length;
      wrapper = mkWrapper({ ...model, config: { ...model.config, removeTile: true, includeTargets: true } }, session);
      wrapper.setProps({
        onSessionChange: jest.fn(newSession => {
          wrapper.setState({ session: newSession })
        })
      });

      wrapper.instance().onRemoveChoice({ id: 'c4', type: 'target', index: 3});

      expect(compact(wrapper.state('session').value).length).toEqual(valuesLength - 1);
    });
  });

  describe('session', () => {
    it('order get restored from session if present', () => {
      let ordering = wrapper.instance().createOrdering();

      expect(ordering.tiles.map(t => t.id)).toEqual(['c1', 'c2', 'c3', 'c4']);

      session = { value: ['c4', 'c2', 'c3', 'c1'] };

      wrapper.setProps({ session });

      ordering = wrapper.instance().createOrdering();

      expect(ordering.tiles.map(t => t.id)).toEqual(['c4', 'c2', 'c3', 'c1']);
    });
  });

  describe('show correct response', () => {
    it('toggle is not visible if correct response is not present', () => {
      session = { value: ['c4', 'c2', 'c3', 'c1'] };
      wrapper = mkWrapper(model, session);

      const toggle = wrapper.find(CorrectAnswerToggle);

      expect(toggle.prop('show')).not.toEqual(true);
      expect(toggle.prop('toggled')).toEqual(false);
    });

    it('toggle is visible if correct response is present', () => {
      session = { value: ['c4', 'c2', 'c3', 'c1'] };
      model.correctResponse = ['c1', 'c4', 'c3', 'c2'];
      wrapper = mkWrapper(model, session);

      const toggle = wrapper.find(CorrectAnswerToggle);

      expect(toggle.prop('show')).toEqual(true);
      expect(toggle.prop('toggled')).toEqual(false);
    });
  });
});
