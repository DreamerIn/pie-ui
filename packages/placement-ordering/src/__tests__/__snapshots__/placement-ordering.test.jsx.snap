// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacementOrdering render snapshot 1`] = `
<div
  className="placementOrdering"
>
  <_default
    onToggle={[Function]}
    toggled={false}
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
  <WithStyles(HTiler)
    instanceId="1"
    onDropChoice={[Function]}
    onRemoveChoice={[Function]}
    tiles={
      Array [
        Object {
          "draggable": true,
          "droppable": true,
          "id": "c1",
          "label": "C1",
          "type": "choice",
        },
        Object {
          "draggable": true,
          "droppable": true,
          "id": "c2",
          "label": "C2",
          "type": "choice",
        },
        Object {
          "draggable": true,
          "droppable": true,
          "id": "c3",
          "label": "C3",
          "type": "choice",
        },
        Object {
          "draggable": true,
          "droppable": true,
          "id": "c4",
          "label": "C4",
          "type": "choice",
        },
      ]
    }
  />
  <br />
  <Feedback />
</div>
`;
